modules/storage/src/main/kotlin/edu/stanford/spezi/modules/storage/key/KeyValueStorage.kt

```diff
@@ -0,0 +1,10 @@
+package edu.stanford.spezi.modules.storage.key
+
+import kotlinx.coroutines.flow.Flow
+
+interface KeyValueStorage {
```

Comment:
I find it great that you added storage support with both data store and shared preferences API 💪 I am noting down some suggestions regarding the API of this module.

1. I would start with a singleton Factory component `KeyValueStorageFactory` with a method, `createStorage(name: String, type: (DataStore/SharePrefs): KeyValueStorage` The factory itself will hold the created storages in a HashMap, if already a storage of the same name and type exists, return that, otherwise create new. It is important to not create more than one data store with the same name, hence the map is important. The `LocalKeyValueStorage` and `EncryptedSharedPreferencesKeyValueStorage` can be marked internal and no need for `@Inject` constructor, since it will be created inside the factory. Extension property datastore on the companion object can move in the class level.

Why a factory? Well, key value storages keep simple values or settings which might also be clustered by feature, hence it makes sense to support creation of different files. The factory will prefix the filenames with `edu.stanford.spezi`, e.g. `APP_SETTINGS` filename will result into `edu.stanford.spezi.APP_SETTINGS`. This allows having all files sorted / grouped in device explorer.

2. Instead of `PreferenceKey<T>` sealed class, I would propose explicit methods for supported types, e.g. readInt, putInt. In the LocalKeyValueStorage implementation, you can use `intPreferencesKey(key)` as an implementation detail, and the exposed `key` of all method signatures will be of type `String`:

```kotlin
suspend fun putInt(key: String, value: Int) {
    dataStore.edit { preferences ->
        preferences[intPreferencesKey(key)] = value
    }
}

suspend fun readInt(key: String): Int? {
    return preferences()
        .map { preferences ->
            preferences[intPreferencesKey(key)]
        }.firstOrNull()
}

private fun preferences() = dataStore.data.catch { emit(emptyPreferences()) }
```

As writing and reading to storages is a pretty common use cases, I think having explicit methods simplifies the API in comparsion to the generic PreferenceKey. We can then create provide default hilt dependencies with a given file name for data store and shared preferences implementation, also enable creation of any arbitrary file name via the factory.

Reply:
Thanks for the feedback. 

I like the first point as it offers more flexibility. 

I am unsure about the second point. The _SharedPreferences_ Library uses an approach that I don't particularly like. I therefore used the more advanced approach of the _DataStore_ implementation and took this as inspiration for the interface which is used by both implementations. I have adapted it so that the _SharedPreferences_ implementation also uses the same interface. I implemented missing features, such as _ByteArray_, custom so that the respective implementations can be exchanged if necessary.

With the data store implementation you can simply create objects like:
`private object PreferencesKeys {
  val SHOW_COMPLETED = booleanPreferencesKey("show_completed")
  val USER_NAME = stringPreferencesKey("user_name") 
}`

and than use the generic method to retrieve any of them. 

So if you have a list of preferences, which I think can happen quickly, you can just dynamically load them all out here, whereas if you have a separate method for each data type, you always have to pick the right method or you would probably write your own class that abstracts this logic, which would actually bring you back to this approach, right?

Reply:
I see 👍 I am already in favor of supporting both APIs as I mentioned. Even though shared preferences is old, it is still massively used from tons of projects due to its simple and reliable API. You can have a look at this [article](https://medium.com/mobile-innovation-network/sharedpreferences-vs-datastore-choosing-the-right-one-72f0789c956c#:~:text=When%20deciding%20between%20Jetpack%20DataStore,DataStore%20is%20a%20better%20choice.) that summarizes the diffs. From my perspective, considering our use cases we should stick with SharedPrefs only. For more complex types, we could persist json string of the instance. This would mean, we support read/write for primitive types only (no need for byte array), and add extension functions on `KeyValueStorage` for read/write serializable at a given key, which internally will encode / decode to json string. No matter what, it is fine as it is now 👍 

My second point was mainly about API design:

```kotlin
suspend fun handle(storage: LocalKeyValueStorage) {
    val key = "someKey"
    // 1. Option
    val value1 = storage.readDataBlocking(key = PreferenceKey.IntKey(name = key))
    // 2. Option
    val value2 = storage.readInt(key = key)
}
```

I personally find option 2 more lightweight and less boilerplate 😅 However, both options are fine by me 👍

Reply:
Adding json later seems like a good idea.

at least my idea would have been that people use it more like this:

`private object UserData { 
val SHOW_COMPLETED = PreferenceKey.BooleanKey("show_completed") 
val USER_NAME = PreferenceKey.StringKey("user_name")
}`

and than it can be read like:

`keyValueStorage.readDataBlocking(SHOW_COMPLETED)
keyValueStorage.readDataBlocking(USER_NAME)
`